### PR TITLE
feat: implement Multisig wallet with unit tests and .learn.md documentation

### PR DESCRIPTION
**Pull request by Sammmmx**
### Overview
This PR completes the task outlined in issue **#7**. I have implemented a robust Multi-Signature wallet contract that supports both native ETH and ERC20 token transactions, along with governance features for managing signers and thresholds.

### Changes Made
* **Smart Contract:** Created `multisig.sol` with proposal, confirmation, and execution logic.
* **Test Contract:** Created `TempERC20.sol` to cover ERC20 Transaction Tests.
* **Security:** Implemented constructor safeguards (zero-address checks, redundancy checks) and used custom errors for gas efficiency.
* **Testing:** Added a comprehensive Hardhat test suite covering:
    * Deployment constraints.
    * Submission and auto-confirmation logic.
    * Execution flow for ETH and ERC20.
    * Edge cases for unauthorized access and insufficient balances.
* **Documentation:** Created a `.learn.md` file providing a technical deep-dive into the contract mechanisms and usage.
* **Configuration:** Updated `hardhat.config.ts` and project structure to support the new module.

### Related Issue
Fixes #7

### CODE REVIEW

**Comment by Sammmmx** [comment]
Sorry to reopen the PR multiple times, I was trying to figure out why checks failed as I am new in Github it took me few mins to learn about the issue in my branch.

**Review by PetarStoev02 — CHANGES_REQUESTED** [review]
Hey @Sammmmx, great effort on this! The core logic is solid and CI is green ✅. Here's my review — a few things to address before we can merge.

> Sorry to reopen the PR multiple times, I was trying to figure out why checks failed as I am new in Github it took me few mins to learn about the issue in my branch.

No worries at all! For future PRs — open it as a **Draft PR** first (there's a dropdown on the "Create pull request" button). That way CI runs on your branch without requesting reviews. Once everything is green and you're happy with it, click "Ready for review" to open it up. That keeps the PR timeline clean and avoids unnecessary notifications to reviewers.

## 🔴 Must Fix

1. **No reentrancy guard** — `execute()` uses low-level `.call{value}` which forwards all gas. A malicious recipient could re-enter `confirm()` or `Submit()` on other pending transactions. Add OpenZeppelin's `ReentrancyGuard`.

2. **`transferERC20` uses `.transfer()` instead of `SafeERC20`** — Tokens like USDT don't return a `bool`, so this will revert on them. Use `SafeERC20.safeTransfer()`.

3. **No confirmation revocation** — Once a signer confirms, it's permanent. A `revoke(uint256 _txId)` function is standard in multisig implementations.

4. **No transaction cancellation** — Malicious/erroneous pending transactions live forever. Add a mechanism to cancel.

5. **`Submit()` callable by anyone** — Anyone can spam proposals. Restrict to signers only.

6. **`ADD_SIGNER` doesn't check `address(0)`** — A proposal to add `address(0)` as a signer can be executed, breaking the signer system.

## 🟡 Should Fix

7. **Constructor O(n²) duplicate check** — Use the `isSigner` mapping instead of nested loops for O(n).
8. **`REMOVE_SIGNER` error reports wrong address** — `NotSigner(msg.sender)` should be `NotSigner(_address)`.
9. **Stale confirmations after signer removal** — If a signer is removed after confirming, their vote still counts. Could allow execution with approvals from non-signers.
10. **No event on `receive()`** — Add a `Deposit` event for off-chain tracking.
11. **`README.md` is default Hardhat boilerplate** — References `Lock.ts`, not the multisig. Update or remove it.
12. **`dotenv` imported in `package.json` but never used** — `hardhat.config.ts` never calls `require('dotenv').config()`.
13. **`async` on `describe` callbacks** — Mocha doesn't await these. Remove `async` from `describe` blocks.

## 🟢 Nice to Have

- Follow Solidity naming conventions (`multisig` → `Multisig`, `Submit` → `submit`, etc.)
- Add NatSpec documentation (`@notice`, `@dev`, `@param`)
- Use named constants for enum values in tests instead of magic numbers (`1`, `2`, `3`...)
- Add more test coverage: reentrancy, stale confirmations, event assertions, `receive()`, threshold edge cases
- Move `TempERC20.sol` from `src/` to a test directory

Overall this is heading in the right direction. Fix the critical items and this will be ready to go 🚀

**Comment by Sammmmx** [comment]
Hello @PetarStoev02, I have addressed your request for necessary changes. Requesting you to review it again, hoping I have fulfilled the Multi Signature Wallet Standards.

File: evm/multisig/src/multisig.sol [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        isSigner[_signer] = false;
+        for (uint256 i = 0; i < signers.length; i++) {
+            if (signers[i] == _signer) {
+                signers[i] = signers[signers.length - 1];
+                signers.pop();
+                break;
+            }
+        }
+        if (threshold > signers.length) {
+            threshold = signers.length < 2 ? 2 : signers.length;
```
Comment (PetarStoev02) [inline evm/multisig/src/multisig.sol]: This can permanently brick the wallet. With 2 signers and threshold=2, removing one drops signers to 1, but this line sets threshold back to 2. Result: threshold=2, signers=1 — no transaction can ever execute again, funds locked forever.

Add a guard before the removal:

`require(signers.length - 1 >= threshold, "Cannot remove: would brick wallet");`

Or at minimum require(signers.length > 2, ...).

File: evm/multisig/src/multisig.sol [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        hasConfirmed[_txId][msg.sender] = false;
+    }
+
+    /// @notice Cancel a transaction
+    /// @param _txId Transaction ID
+    function cancel(uint256 _txId) public onlySigner {
+        require(_txId < transactions.length, TxNonExistent());
+        require(transactions[_txId].state == States.PENDING, TxNotPending());
+        
+        transactions[_txId].state = States.CANCELLED;
```
Comment (PetarStoev02) [inline evm/multisig/src/multisig.sol]: Any single signer can unilaterally cancel any pending transaction, even one that already has threshold confirmations. This is asymmetric with the rest of the design — submitting and executing require M-of-N consensus but cancelling only requires 1-of-N. A rogue signer could grief by cancelling everything.

Consider requiring threshold confirmations for cancellation, or restricting to the original submitter (would need to store submitter in the Transaction struct).

File: evm/multisig/src/multisig.sol [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        }));
+
+    _tryAutoConfirm(txId);
+    emit Submitted(txId, msg.sender, _to, _value);
+    }
+
+
+    /// @notice Confirm a transaction
+    /// @param _txId Transaction ID
+    function confirm(uint256 _txId) public onlySigner {
```
Comment (PetarStoev02) [inline evm/multisig/src/multisig.sol:172]: Missing Confirmed event. There's no event Confirmed(...) declared and this function emits nothing. The test for it at line 381 silently passes due to incorrect await placement. Add:


`event Confirmed(uint256 indexed txId, address indexed signer);`
and emit it at the end of this function.

File: evm/multisig/.learn.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  - Confirms that state updates to `EXECUTED` properly.
+- **Edge Cases:** Tests for insufficient ETH balance and unauthorized confirmation attempts.
+
+## Usage Summary
+
+1. **Deploy:** Provide a list of addresses and a threshold (e.g., `[A, B, C], 2`).
+2. **Fund:** Send ETH or ERC20 tokens to the contract address.
+3. **Propose:** Call `Submit` with the desired action.
+4. **Approve:** Other signers call `confirm(txId)`.
+5. **Finalize:** Once approvals $\ge$ Threshold, anyone calls `execute(txId)`.
```
Comment (PetarStoev02) [inline evm/multisig/.learn.md:110]: This contradicts the code. execute() has the onlySigner modifier, so only signers can call it — not "anyone". Update to: "any signer calls execute(txId)".

File: evm/multisig/test/multisigTest.ts [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+    });
+
+    it("should revert if signer already confirmed", async function () {
+      await multiSig.connect(Signer2).confirm(0);
+      await expect(
+        multiSig.connect(Signer2).confirm(0),
+      ).to.be.revertedWithCustomError(multiSig, "AlreadyConfirmed");
+    });
+
+    it("should emit Confirmed event", async function () {
```
Comment (PetarStoev02) [inline evm/multisig/test/multisigTest.ts:381]: **False positive test.** This uses `expect(await tx).to.emit()` — the `await` is in the wrong place. Chai needs the unawaited promise to intercept events. Correct pattern:
```typescript
await expect(multiSig.connect(Signer2).confirm(0))
  .to.emit(multiSig, "Confirmed")
  .withArgs(0, Signer2.address);
```
Also, the `Confirmed` event needs to be added to the contract first (see comment on `confirm()`).

File: evm/multisig/test/multisigTest.ts [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+      SYMBOL,
+      MAX_SUPPLY,
+    )) as TokenERC20;
+
+    await token.mint(multiSig.target, ethers.parseEther("1000"));
+  });
+
+  // Deployment Tests
+
+  describe("Deployment", async function () {
```
Comment (PetarStoev02) [inline evm/multisig/test/multisigTest.ts]: **Remove `async` from `describe` callbacks.** Mocha doesn't await these — it's a no-op that can mask errors. This applies to all the old `describe` blocks (lines 58, 144, 158, 221, 269, 297, 317, 350, 558, 581, 645, 672, 692, 712). Quick find-and-replace: `describe("...", async function ()` → `describe("...", function ()`

File: evm/multisig/src/multisig.sol [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    /// @notice Transaction states
+    enum States { NONE, PENDING, EXECUTED, CANCELLED }
+
+    /// @notice Transaction structure
+    /// @param to Target address
+    /// @param token ERC20 token address
+    /// @param value ETH or token amount
+    /// @param txType Transaction type
+    /// @param state Transaction state
+    /// @param confirmations Number of confirmations
```
Comment (PetarStoev02) [inline evm/multisig/src/multisig.sol]: **Ghost NatSpec.** The `Transaction` struct has no `confirmations` field — this documents something that doesn't exist. Remove this line.

**Review by PetarStoev02 — CHANGES_REQUESTED** [review]
Great progress @Sammmmx — most of the original feedback is addressed nicely. Two new issues to fix: the `_removeSigner` wallet bricking bug (critical) and the unilateral `cancel()`. The rest are small. You're close 🚀

File: evm/multisig/src/multisig.sol (hunk 2 above)
Comment (Sammmmx) [reply evm/multisig/src/multisig.sol]: I have restricted it to original submitter.

File: evm/multisig/src/multisig.sol (hunk 3 above)
Comment (Sammmmx) [reply evm/multisig/src/multisig.sol:172]: Added event Confirmed

File: evm/multisig/.learn.md (hunk 4 above)
Comment (Sammmmx) [reply evm/multisig/.learn.md:110]: removed onlySigner() modifier from execute()

File: evm/multisig/test/multisigTest.ts (hunk 5 above)
Comment (Sammmmx) [reply evm/multisig/test/multisigTest.ts:381]: adjusted await inside expect()

File: evm/multisig/test/multisigTest.ts (hunk 6 above)
Comment (Sammmmx) [reply evm/multisig/test/multisigTest.ts]: removed async from all describe blocks

File: evm/multisig/src/multisig.sol (hunk 7 above)
Comment (Sammmmx) [reply evm/multisig/src/multisig.sol]: Removed /// @param Confirmations and replaced with /// @param Porposer

**Review by PetarStoev02 — APPROVED** [review]
Thanks for the great work @Sammmmx! I pushed a small cleanup commit with the remaining fixes (brick-wallet guard, cancelled event, docs sync, gas optimizations). All 61 tests passing ✅

Feel free to pick up another issue from the board when you're ready!
